pkg/db/db.go

```diff
@@ -757,6 +805,216 @@ func (h *DBHandler) RunCustomMigrationAllTables(ctx context.Context, getAllAppsF
+		}
+		return &EnvironmentLock{
+			EslVersion: row.EslVersion,
+			Created:    row.Created,
+			LockID:     row.LockID,
+			Env:        row.Env,
+			Metadata:   resultJson,
+		}, nil
+	}
+	return nil, nil // no rows, but also no error
+
+}
+
+func (h *DBHandler) DBWriteEnvironmentLock(ctx context.Context, tx *sql.Tx, lockID, environment, message, authorName, authorEmail string) error {
+	if h == nil {
+		return nil
+	}
+	if tx == nil {
+		return fmt.Errorf("DBWriteEnvironmentLock: no transaction provided")
+	}
+	span, _ := tracer.StartSpanFromContext(ctx, "DBWriteEnvironmentLock")
+	defer span.Finish()
+
+	var previousVersion EslId
+
+	//See if there is an existing lock with the same lock id in this environment. If it exists, just add a +1 to the eslversion
+	existingEnvLock, err := h.DBSelectEnvironmentLock(ctx, tx, environment, lockID)
+
+	if err != nil {
+		return fmt.Errorf("Could not obtain existing environment lock: %w\n", err)
+	}
+
+	if existingEnvLock == nil {
+		previousVersion = 0
+	} else {
+		previousVersion = existingEnvLock.EslVersion
+	}
+
+	envLock := EnvironmentLock{
+		EslVersion: 0,
+		LockID:     lockID,
+		Created:    time.Time{},
+		Env:        environment,
+		Metadata: EnvironmentLockMetadata{
+			Message:        message,
+			CreatedByName:  authorName,
+			CreatedByEmail: authorEmail,
+		},
+	}
+	return h.DBWriteEnvironmentLockInternal(ctx, tx, envLock, previousVersion)
+}
+
+func (h *DBHandler) DBWriteEnvironmentLockInternal(ctx context.Context, tx *sql.Tx, envLock EnvironmentLock, previousEslVersion EslId) error {
+	if h == nil {
+		return nil
+	}
+	if tx == nil {
+		return fmt.Errorf("dbWriteEnvironmentLockInternal: no transaction provided")
+	}
+	span, _ := tracer.StartSpanFromContext(ctx, "dbWriteEnvironmentLockInternal")
+	defer span.Finish()
+
+	jsonToInsert, err := json.Marshal(envLock.Metadata)
+	if err != nil {
+		return fmt.Errorf("could not marshal json data: %w", err)
+	}
+
+	insertQuery := h.AdaptQuery(
+		"INSERT INTO environment_locks (eslVersion, created, lockID, envName, metadata) VALUES (?, ?, ?, ?, ?);")
+
+	span.SetTag("query", insertQuery)
+	_, err = tx.Exec(
+		insertQuery,
+		previousEslVersion+1,
+		time.Now(),
+		envLock.LockID,
+		envLock.Env,
+		jsonToInsert)
+
+	if err != nil {
+		return fmt.Errorf("could not write environment lock into DB. Error: %w\n", err)
+	}
+	return nil
+}
+
+func (h *DBHandler) DBSelectEnvLocks(ctx context.Context, tx *sql.Tx, environmentName string) ([]EnvironmentLock, error) {
+	if h == nil {
+		return nil, nil
+	}
+	if tx == nil {
+		return nil, fmt.Errorf("DBSelectEnvLocks: no transaction provided")
+	}
+	span, _ := tracer.StartSpanFromContext(ctx, "DBSelectEnvLocks")
+	defer span.Finish()
+
+	selectQuery := h.AdaptQuery(
+		fmt.Sprintf(
+			"SELECT eslVersion, created, lockID, envName, metadata" +
+				" FROM environment_locks " +
+				" WHERE envName=? " +
+				" ORDER BY eslVersion DESC " +
```

Comment:
this will not get you the last 25 locks, it will give you the last 25 changes in locks.

Reply:
Isn't this the same thing? Recorded changes  to the environment_locks table are just insertions. Deletions remove the locks from the table altogether.  In my eyes this query returns the last 25 locks that were created on a certain environment (and still exist)